### PR TITLE
Add null_value_pattern and null_empty_string option for LabeledTSVParser, CSVParser, TSVParser.

### PR DESCRIPTION
**Pull request by toyama0919**
Hello.

Cannot express NULL in LTSV.

ltsv library for Ruby blank to NULL.
https://github.com/condor/ltsv/blob/master/lib/ltsv.rb#L97

I use LTSV with apache.
Apache log is If value not exists with hyphen('-').

I think hyphens to null, any of the following methods.

1. LabeledTSVParser add parameter with @null_value
※this PR

settings example.
```
<source>
  type tail
  path /var/log/httpd/access_log
  format ltsv
  time_key time
  null_value -
  time_format %d/%b/%Y:%H:%M:%S %z
  tag td.apache.access
  pos_file /var/log/td-agent/apache_access.pos
</source>
```

2. specifications similar to the Ruby library(blank to null)

3. Make your own filter plugin.

Which one is good?

### CODE REVIEW

**Comment by frsyuki** [comment]
I think 1. is good and it is good enough for now.


I may want to use using 1. and 2. together occasionally. So, alternative idea is to add `null_value_pattern: pattern` option instead of `null_value` to accept a regexp, or add `null_empty_string: true` option in addition to `null_value`.

**Comment by frsyuki** [comment]
I think this PR is simple enough and good to merge. Any concerns, @repeatedly?

**Comment by toyama0919** [comment]
@frsyuki
Thank you for your reply!

>I may want to use using 1. and 2. together occasionally. So, alternative idea is to add null_value_pattern: pattern option instead of null_value to accept a regexp, or add null_empty_string: true option in addition to null_value.

I see.

Please wait a moment...

**Comment by toyama0919** [comment]
Thank you for advice!

I add null_value_pattern and null_empty_string option for LabeledTSVParser.

settings example.
```
<source>
  type tail
  path /var/log/httpd/access_log
  format ltsv
  time_key time
  null_value_pattern ^(-|null|NULL)$
  null_empty_string true
  time_format %d/%b/%Y:%H:%M:%S %z
  tag td.apache.access
  pos_file /var/log/td-agent/apache_access.pos
</source>
```

## null_value_pattern
* default is nil.
* regular expression syntax.

## null_empty_string
* default is false.
* empty string to null.
* [this](https://github.com/condor/ltsv/blob/master/lib/ltsv.rb#L97) is same behavior.

File: lib/fluent/parser.rb [hunk 1]
```diff
@@ -381,6 +383,12 @@ def parse(text)
         text.split(delimiter).each do |pair|
           key, value = pair.split(label_delimiter, 2)
           @keys.push(key)
+          if null_empty_string
+            value = (value == '') ? nil : value
+          end
+          if null_value_pattern
+            value = (value =~ /#{null_value_pattern}/) ? nil : value
```
Comment (repeatedly) [inline lib/fluent/parser.rb]: Creating regexp object at each record is high cost.
It should be done in configure method.

**Comment by sonots** [comment]
Why don't we add these options for super class (ValuesParser)?

File: lib/fluent/parser.rb (hunk 1 above)
Comment (toyama-h) [inline lib/fluent/parser.rb]: thank you!
I'll fix it.

**Comment by toyama-h** [comment]
@sonots
thank you!
Oh, ValuesParser definition, affects CSVParser and TSVParser.
Null in regular expression definition, I think Good idea!
But empty to NULL casting is LTSV specification.

@repeatedly
I think no problem, Is it OK?

**Comment by toyama-h** [comment]
Certainly, in the format like below with MySQL restore

```
"1";"2";"3";"4";NULL
```

**Comment by sonots** [comment]
> empty to NULL casting is LTSV specification.

Is it official LTSV specification? I thought it is just a custom.

**Comment by toyama0919** [comment]
>Is it official LTSV specification? I thought it is just a custom.

That's right!
No such process in other languages.

### java
https://github.com/making/ltsv4j/blob/master/src/main/java/am/ik/ltsv4j/LTSVParser.java

### python
https://github.com/hekyou/python-ltsv/blob/master/ltsv/_reader.py

**Comment by toyama0919** [comment]
>Why don't we add these options for super class (ValuesParser)?
I'll fix fix in this way.
Please wait a moment.

**Comment by toyama-h** [comment]
fixed.
add null_value_pattern and null_empty_string option, support format is tsv,csv,ltsv.

File: lib/fluent/parser.rb [hunk 2]
```diff
@@ -308,11 +310,16 @@ def configure(conf)
         end
 
         @time_parser = TimeParser.new(@time_format)
+
+        if @null_value_pattern
+          @null_value_pattern = /#{@null_value_pattern}/
```
Comment (sonots) [inline lib/fluent/parser.rb]: `Regexp.new` is better because users do not need to care of escaping `/` with it.

File: lib/fluent/parser.rb [hunk 3]
```diff
@@ -345,6 +352,16 @@ def convert_field_type!(record)
           end
         }
       end
+
+      def convert_value_to_nil(value)
+        if @null_empty_string
+          value = (value == '') ? nil : value
+        end
+        if @null_value_pattern
```
Comment (sonots) [inline lib/fluent/parser.rb]: `if value and @null_value_pattern` is more clear to tell returning nil when `nil =~ @null_value_pattern`

File: lib/fluent/parser.rb [hunk 4]
```diff
@@ -345,6 +352,16 @@ def convert_field_type!(record)
...
+          value = (value =~ @null_value_pattern) ? nil : value
```
Comment (sonots) [inline lib/fluent/parser.rb]: String scrub is needed? cf. http://blog.livedoor.jp/sonots/archives/34702351.html

**Comment by sonots** [comment]
reviewed

File: lib/fluent/parser.rb (hunk 4 above)
Comment (sonots) [inline lib/fluent/parser.rb]: I usually does String#scrub in my plugin as https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/filter_grep.rb#L73-L83. 
I am not sure for the case of fluentd itself. What do you think? > @repeatedly

**Comment by toyama-h** [comment]
thank you for review.
I 'll fix it.

File: lib/fluent/parser.rb (hunk 4 above)
Comment (toyama0919) [inline lib/fluent/parser.rb]: sorry.

I think this private method add ValuesParser, Is it OK?

```
      def match(regexp, string)
        begin
          return regexp.match(string)
        rescue ArgumentError => e
          raise e unless e.message.index("invalid byte sequence in".freeze).zero?
          log.info "invalid byte sequence is replaced in `#{string}`"
          string = string.scrub('?')
          retry
        end
        return true
      end

```

**Comment by toyama0919** [comment]
Apply this review.
scrub Referenced filter-grep-plugin.

thanks.

File: lib/fluent/parser.rb (hunk 4 above)
Comment (sonots) [inline lib/fluent/parser.rb]: Creating Fluent::StringUtil should be better.

Comment (toyama0919) [inline lib/fluent/parser.rb]: Ah OK!
thanks.

**Comment by toyama0919** [comment]
added Fluent::StringUtil, filter-grep-plugin and this feature common procesing.
thanks.

**Comment by sonots** [comment]
Looks good to me. Any concerns? > @repeatedly

**Comment by repeatedly** [comment]
Nothing.

**Comment by sonots** [comment]
Merged!

**Comment by toyama-h** [comment]
@repeatedly @sonots @frsyuki thank you for many advise! :blush:

**Comment by repeatedly** [comment]
Released v0.12.16!

**Comment by toyama-h** [comment]
:blush:

**Comment by tagomoris** [comment]
I found that `log` is not defined in `Fluent::StringUtil.match_regexp`. So this rescue block will raise NameError every time.

**Comment by toyama0919** [comment]
@tagomoris Sorry and Thanks. I will check it.

**Pull request merged**